src/glib/mine/svm.h

```diff
@@ -13,376 +13,640 @@
+    TSvmParam(TInt _Type, TInt _Kernel, TFlt _Cost, TFlt _Unbalance, TFlt _Gamma, TInt _Degree, TFlt _Coef0,  TFlt _CacheSize, TFlt _Eps, TFlt _P, TFlt _Nu) : 
+        Type(_Type), Kernel(_Kernel), Degree(_Degree), Cost(_Cost), Unbalance(_Unbalance), Gamma(_Gamma), Nu(_Nu), Coef0(_Coef0), Eps(_Eps), CacheSize(_CacheSize), P(_P), Shrink(0), Prob(0)  {  } //fill with default values
+    
+    TSvmParam(TSIn& SIn):
+      Type(TInt(SIn)),
+      Kernel(TInt(SIn)),
+      Degree(TInt(SIn)),
+      Cost(TFlt(SIn)),
+      Unbalance(TFlt(SIn)),
+      Gamma(TFlt(SIn)),
+      Nu(TFlt(SIn)),
+      Coef0(TFlt(SIn)),
+      Eps(TFlt(SIn)),
+      CacheSize(TFlt(SIn)),
+      P(TFlt(SIn)),
+      Shrink(TInt(SIn)),
+      Prob(TInt(SIn)) {  }
+    
+    void Save(TSOut& SOut) const {
+      TInt(Type).Save(SOut);
+      TInt(Kernel).Save(SOut);
+      TInt(Degree).Save(SOut);
+      TFlt(Cost).Save(SOut);
+      TFlt(Unbalance).Save(SOut);
+      TFlt(Gamma).Save(SOut);
+      TFlt(Nu).Save(SOut);
+      TFlt(Coef0).Save(SOut);
+      TFlt(Eps).Save(SOut);
+      TFlt(CacheSize).Save(SOut);
+      TFlt(P).Save(SOut);
+      TInt(Shrink).Save(SOut);
+      TInt(Prob).Save(SOut);
+    }
+    
+    svm_parameter_t GetParamStruct() const {//returns svm_parameter_t for LIBSVM
+      svm_parameter_t svm_parameter;
+      svm_parameter.svm_type = this->Type;//default
+      svm_parameter.kernel_type = this->Kernel;//default
+      svm_parameter.degree = this->Degree;
+      svm_parameter.C = this->Cost;
+      svm_parameter.gamma = this->Gamma;
+      svm_parameter.nu = this->Nu;
+      svm_parameter.coef0 = this->Coef0;
+      svm_parameter.nr_weight = 2;
+      svm_parameter.weight_label = (int *)malloc(2 * sizeof(int)); // deleted in svm_destroy_param
+      svm_parameter.weight_label[0] = -1;
+      svm_parameter.weight_label[1] = 1;
+      svm_parameter.weight = (double *)malloc(2 * sizeof(double));  // deleted in svm_destroy_param
+      svm_parameter.weight[0] = 1;
+      svm_parameter.weight[1] = Unbalance;
+      svm_parameter.cache_size = this->CacheSize;
+      svm_parameter.eps = this->Eps;
+      svm_parameter.p = this->P; // not needed but it has to be positive as it is checked
+      svm_parameter.shrinking = this->Shrink;
+      svm_parameter.probability = this->Prob;
+      return svm_parameter;
+    }
+};//end TSvmParam
 
 /// Linear model
 class TLinModel {
 private:
     TFltV WgtV;
     TFlt Bias;
+    TSvmParam Param;
+    TInt NClass; //number of classes
+    TInt NSV; //number of support vectors
+    TFltVV SV; //support vectors
+    TFltVV Coef; //coefficients for support vectors
+    TFltV Rho; //bias terms
+    TFltV ProbA;
+    TFltV ProbB;
+    TIntV SVIndices; //indices of support vectors in data
+    TIntV Label; // for classification only
+    TIntV nSV; // for classification only
+    TInt Free;
```

Comment:
How many of the members above we actually need for doing predictions? In case we have linear kernel, we just need weights and bias. If we have non-linera, we need support vectors, coefficients and bias + any kernel parameters. As it is now we keep all the training parameters in here as well, which are only relevant when we fit.

Looking at the code below, looks like we need them to call libsvm predict functions. But what if we would do for kernels same thing as we did for linear weights, where we do our own predictions and not rely anymore on libsvm? It would be faster and cleaner and remove then need to drag around their luggage if all we need is few support vectors and their coefficients.